docs/2.10.0/docs/concepts/ledger-model/upgrade-model-in-depth.rst

```diff
@@ -0,0 +1,1583 @@
+-  If the LF versions of both *p1* and *p2* support upgrades then
+      package *p2* is a valid upgrade of package *p1*, else *p2* and
+      *p1* are the exact same package.
+
+It is worth noting that even when *t2* upgrades *t1*, *r2* only upgrades
+*r1* provided that package *p2* is a valid upgrade of package *p1* as a
+whole.
+
+.. _examples-9:
+
+**Examples**
+
+In these examples we assume the existence of packages ``q-1.0.0`` and
+``q-2.0.0`` with LF version 1.16, and that the latter is a valid upgrade of
+the former.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+
+   * - In ``q-1.0.0``:
+     - In ``q-2.0.0``:
+
+   * - .. code-block:: daml
+
+            module Dep where
+     
+            data U = C1
+            data V = V
+     
+     - :.. code-block:: daml
+     
+            module Dep where
+     
+            data U = C1 | C2
+            data V = V
+     
+Then below, the module on the right is a valid upgrade of the module on
+the left.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+
+   * - .. code-block:: daml
+
+            module Main where
+     
+            -- imported from q-1.0.0
+            import qualified Dep
+     
+            data T = T Dep.U
+     
+     - .. code-block:: daml
+
+            module Main where
+     
+            -- imported from q-2.0.0
+            import qualified Dep
+     
+            data T = T Dep.U
+     
+However below, the module on the right is **not** a valid upgrade of the
+module on the left because ``Dep.V`` on the right belongs to package ``q-1.0.0``
+which is not a valid upgrade of package ``p-2.0.0``, even though the two
+definitions of ``V`` are the same.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+
+   * - .. code-block:: daml
+
+            module Main where
+     
+            -- imported from q-2.0.0
+            import qualified Dep
+     
+            data T = T Dep.V
+
+     - .. code-block:: daml
+
+            module Main where
+     
+            -- imported from q-1.0.0
+            import qualified Dep
+     
+            data T = T Dep.V
+
+Suppose now that q-1.0.0 and q-2.0.0 are both compiled to LF version
+1.15 (which does not support upgrades). Then below, the module on the
+right is **not** a valid upgrade of the module on the left because the
+references to U on each side resolve to packages with different IDs.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+
+   * - .. code-block:: daml
+
+            module Main where
+     
+            -- imported from q-1.0.0
+            import qualified Dep
+     
+            data T = T Dep.U
+     
+     - .. code-block:: daml
+
+            module Main where
+     
+            -- imported from q-2.0.0
+            import qualified Dep
+     
+            data T = T Dep.U
+
+Data Types - Parameterized Data Types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Parameterized data types are considered serializable for the purpose of
```

Comment:
Found this statement a bit strange. This is not only for the purpose of upgrade validation, but for all purpose.